README.md
# Avaliageo


Um sistema para avaliar referências geográficas presentes em notícias a partir de um portal de notícias.


Este sistema foi desenvolvido como trabalho de conclusão de curso no curso de Engenharia de Computação na Universidade Federal de Ouro Preto.



preProssedNews/news.txt
[["https://g1.globo.com/economia/noticia/2019/03/28/carrefour-cortara-mais-de-1-mil-empregos-na-franca-diz-sindicato.ghtml", ["<div class=\"collapse\" id=\"quest_0\"> <div class=\"well\"> <div class=\"row\"> <div style=\"margin: 0 auto;\" class=\"text-center\"><div class=\"col-md-1\">Topônimo:</div> <div class=\"col-md-6\"><select class=\"form-control style_select\" name=\"top_data_0\" id=\"top_data_0\"><option value=\"0 2992090 Montreuil PPL /Ilha---de---França FR\">Montreuil (Cidade /Ilha de França / França)</option><option value=\"1 2992061 Montreuil PPLA3 /Hauts-de-France FR\">Montreuil (Cidade /Hauts-de-France / França)</option><option value=\"2 3033415 Berck PPL /Hauts-de-France FR\">Berck (Cidade /Hauts-de-France / França)</option><option value=\"3 2992060 Montreuil-sur-Thérain PPL /Hauts-de-France FR\">Montreuil-sur-Thérain (Cidade /Hauts-de-France / França)</option><option value=\"4 3008387 La Madelaine-sous-Montreuil PPL /Hauts-de-France FR\">La Madelaine-sous-Montreuil (Cidade /Hauts-de-France / França)</option><option value=\"5 0000000 Nenhuma das Alternativas\">Nenhuma das Alternativas</option> <option value=\"6 0000000 Não é um lugar\">Não é um lugar</option> <option value=\"7 0000000 Não Sei\">Não Sei</option> <option value=\"8 0000000 -- None --\" selected>-- None --</option></select> </div> <div class=\"col-md-2\">  Certeza da Resposta :</div> <div class=\"col-md-1\"><select class=\"form-control style_select\" name=\"confi_data_0\"><option value=\"5\"> 100 %</option> <option value=\"4\"> 75 % </option> <option value=\"3\"> 50 % </option> <option value=\"2\"> 25 % </option> <option value=\"1\"> 0 % </option><option value=\"None\" selected>None</option>  </select> </div> <div class=\"col-md-2\" style=\"justify-items: center;\"><button id=\"confirm_btn_id_0\" class=\"btn btn-success style=\"margin-top:5px;\" type=\"button\" data-toggle=\"collapse\" data-target=\"#quest_0 \" aria-expanded=\"false\" aria-controls=quest_0 onclick=\"submitButtonStyle(this)\" >Confirmar</button> </div></div></div></div></div><div class=\"collapse\" id=\"quest_1\"> <div class=\"well\"> <div class=\"row\"> <div style=\"margin: 0 auto;\" class=\"text-center\"><div class=\"col-md-1\">Topônimo:</div> <div class=\"col-md-6\"><select class=\"form-control style_select\" name=\"top_data_1\" id=\"top_data_1\"><option value=\"0 3017382 França PCLI / FR\">França (País / França)</option><option value=\"1 3570675 Fort-de-France PPLC /Martinique MQ\">Fort-de-France (Cidade /Martinique / Martinica)</option><option value=\"2 2988507 Paris PPLC /Ilha---de---França FR\">Paris (Cidade /Ilha de França / França)</option><option value=\"3 4030656 Polinésia Francesa PCLD / PF\">Polinésia Francesa (Entidade Política Dependente / Polinésia Francesa)</option><option value=\"4 3012874 Ilha de França ADM1 /Ilha---de---França FR\">Ilha de França (Estado /Ilha de França / França)</option><option value=\"5 0000000 Nenhuma das Alternativas\">Nenhuma das Alternativas</option> <option value=\"6 0000000 Não é um lugar\">Não é um lugar</option> <option value=\"7 0000000 Não Sei\">Não Sei</option> <option value=\"8 0000000 -- None --\" selected>-- None --</option></select> </div> <div class=\"col-md-2\">  Certeza da Resposta :</div> <div class=\"col-md-1\"><select class=\"form-control style_select\" name=\"confi_data_1\"><option value=\"5\"> 100 %</option> <option value=\"4\"> 75 % </option> <option value=\"3\"> 50 % </option> <option value=\"2\"> 25 % </option> <option value=\"1\"> 0 % </option><option value=\"None\" selected>None</option>  </select> </div> <div class=\"col-md-2\" style=\"justify-items: center;\"><button id=\"confirm_btn_id_1\" class=\"btn btn-success style=\"margin-top:5px;\" type=\"button\" data-toggle=\"collapse\" data-target=\"#quest_1 \" aria-expanded=\"false\" aria-controls=quest_1 onclick=\"submitButtonStyle(this)\" >Confirmar</button> </div></div></div></div></div><div class=\"collapse\" id=\"quest_2\"> <div class=\"well\"> <div class=\"row\"> <div style=\"margin: 0 auto;\" class=\"text-center\"><div class=\"col-md-1\">Topônimo:</div> <div class=\"col-md-6\"><select class=\"form-control style_select\" name=\"top_data_2\" id=\"top_data_2\"><option value=\"0 2849464 Regis-Breitingen PPL /Sachsen DE\">Regis-Breitingen (Cidade /Sachsen / Alemanha)</option><option value=\"1 6323178 Lebon Régis ADM2 /Santa---Catarina BR\">Lebon Régis (Distrito do Estado /Santa Catarina / Brasil)</option><option value=\"2 2655262 Bognor Regis PPL /Inglaterra GB\">Bognor Regis (Cidade /Inglaterra / Grã-Bretanha (Reino Unido, UK))</option><option value=\"3 3458671 Lebon Régis PPL /Santa---Catarina BR\">Lebon Régis (Cidade /Santa Catarina / Brasil)</option><option value=\"4 6320391 Pedro Régis ADM2 /Paraíba BR\">Pedro Régis (Distrito do Estado /Paraíba / Brasil)</option><option value=\"5 0000000 Nenhuma das Alternativas\">Nenhuma das Alternativas</option> <option value=\"6 0000000 Não é um lugar\">Não é um lugar</option> <option value=\"7 0000000 Não Sei\">Não Sei</option> <option value=\"8 0000000 -- None --\" selected>-- None --</option></select> </div> <div class=\"col-md-2\">  Certeza da Resposta :</div> <div class=\"col-md-1\"><select class=\"form-control style_select\" name=\"confi_data_2\"><option value=\"5\"> 100 %</option> <option value=\"4\"> 75 % </option> <option value=\"3\"> 50 % </option> <option value=\"2\"> 25 % </option> <option value=\"1\"> 0 % </option><option value=\"None\" selected>None</option>  </select> </div> <div class=\"col-md-2\" style=\"justify-items: center;\"><button id=\"confirm_btn_id_2\" class=\"btn btn-success style=\"margin-top:5px;\" type=\"button\" data-toggle=\"collapse\" data-target=\"#quest_2 \" aria-expanded=\"false\" aria-controls=quest_2 onclick=\"submitButtonStyle(this)\" >Confirmar</button> </div></div></div></div></div><div class=\"collapse\" id=\"quest_3\"> <div class=\"well\"> <div class=\"row\"> <div style=\"margin: 0 auto;\" class=\"text-center\"><div class=\"col-md-1\">Topônimo:</di

files.py
# -*- coding: utf-	8 -*-
import json
import collections
import statistics
import numpy 
#import ctypes
#import ast

def getNewsDataFromUrls():
    with open("preProssedNews/news.txt","r") as file:
        string = file.read();

    return json.loads(string)

def dataFromUser(title, data):
	path = title.replace(" ","_")
	full_path = "bases/"+path+".txt"
	with open(full_path,"a+") as file:
		json_str = json.dumps(data, ensure_ascii=False).encode('utf8')+"@"
		file.write(json_str)
	return json_str


def readFile(title):
	path = title.replace(" ","-")
	full_path = "bases/"+path+".txt"

	with open(full_path,"r") as file:
		string = file.read();
	return string

def printDataFromUser(data):
	for i in data:
		print i[0]
		for j in i[1]:
			print j[0], j[1]

def createDataList(string):
	data = string.split('@')
	del data[-1]
	l = []
	for i in data:
		l.append(eval(i))

	return l

def keyValue(item):
	return item[1]

def lprint(l):
	for i in l:
		print i

# Create a function called "chunks" with two arguments, l and n:
def chunks(l, n):
    # For item i in a range that is a length of l,
    for i in range(0, len(l), n):
        # Create an index range for l of n items:
        yield l[i:i+n]


def CronbachAlpha(itemscores):
    itemscores = numpy.asarray(itemscores)
    itemvars = itemscores.var(axis=1, ddof=1)
    tscores = itemscores.sum(axis=0)
    nitems = len(itemscores)

    return nitems / (nitems-1.) * (1 - itemvars.sum() / tscores.var(ddof=1))

def calculateAlpha(title):
	txt_l =  readFile(title)
	file_data_list = createDataList(txt_l)
	
	aux_l = []
	news = []
	hash_tops = []
	number_of_tops = len(file_data_list[0][0][1])

	for i in file_data_list:
		for j in i:
			for k in j[1]:
				aux_l.append(k[0])
	
	news = list(chunks(aux_l,number_of_tops))
	for i in news:
		hash_tops.append(list(map(lambda x: int(x.split()[0]), i)))

	return CronbachAlpha(hash_tops)

def split(a, n):
    k, m = divmod(len(a), n)
    return (a[i * k + min(i, m):(i + 1) * k + min(i + 1, m)] for i in xrange(n))

def createDataList(string):
	data = string.split('@')
	del data[-1]
	l = []
	for i in data:
		l.append(eval(i))

	return l

def countUsers(title):
	read_f = readFile(title)
	file_data_list = createDataList(read_f)
	return len(file_data_list)

def groupData(title):
	txt_l =  readFile(title)
	file_data_list = createDataList(txt_l)
	number_of_tops = len(file_data_list[0][0][1])
	nl = []
	aux = []
	count = 0

	while count < number_of_tops:
		for i in file_data_list:
			aux.append(i[0][1][count])

		count = count + 1
	
	nl = list(split(aux, number_of_tops))

	return nl

def processData(title,url):
	
	minimum_value_to_accept_news = 2
	std_dev_to_accept = 10

	nl = groupData(title)
	l_aux = []
	aux_l = []
	mean_stddev = []
	info_top_to_carry = ""
	

	if len(nl[0]) < minimum_value_to_accept_news:
		return False

	# run all tops already grouped
	for i in nl:
		# print "-------"
		del l_aux[:]
		del aux_l[:]

		# sort most common top in all votes from users
		for j in i:
			l_aux.append(j[0])
		ctr = collections.Counter(l_aux).most_common(8)
		# print "Most Popular : ", ctr[0][1],"2nd : ", ctr[1][1]

		#verify if there are more then one 'most popular'
		if len(ctr) > 1:
			if ctr[0][1] <= ctr[1][1]:
				# print "return False"
				return False

		# run in all votes for a spacific top and calculate some descriptive statistics
		for j in i:
			if j[0] == ctr[0][0]:
				info_top_to_carry =  j[0]
				aux_l.append(int(j[1]))
		
		mea = statistics.mean(aux_l)
		std_dev = statistics.stdev(aux_l)

		# verify if std_dev is acceptable
		if std_dev > std_dev_to_accept:
			return False

		mean_stddev.append((info_top_to_carry,mea,std_dev))

	# print mean_stddev
	# VOLTAR AQUIIIIII ---------------------------------------------------------
 	return (1.0, mean_stddev)
	

def generateNewsJsonFiles(processData, tops, url,title):

	pd = processData[1]
	number_of_decimal_cases = 4

 	json_news = {
		"title" : "",
		"url" : "",
		"cronbach" : "",
		"number_of_voters" : 0,
		"toponyms" : []
	}

	json_news["title"] = title
	json_news["url"] = url
	json_news["cronbach"] = round(processData[0],4)
	json_news["number_of_voters"] = countUsers(title)

	js = {}

	for i in range(len(tops)):
		js = {
			"top_find_on_new" : "",
			"top_selected_by_user" : "" ,
			"toponym_geonamesId" : 0,
			"mean_confiability" : 0,
			"std_deviation" : 0
			
		}
		js["top_find_on_new"] = tops[i]
		
		if len(pd[i][0].split()[-2]) != 2:
			js["top_selected_by_user"] = " ".join(pd[i][0].split()[2:-2])
		else:
			js["top_selected_by_user"] = " ".join(pd[i][0].split()[2:-2])

		js["toponym_geonamesId"] = pd[i][0].split()[1]
		js["mean_confiability"] = round(float(pd[i][1]),number_of_decimal_cases)
		js["std_deviation"] = round(float(pd[i][2]),number_of_decimal_cases)

		json_news['toponyms'].append(js)

	# , ensure_ascii=False).encode('utf8'
	closed_news = json.dumps(json_news, sort_keys=False,indent=4, ensure_ascii=False).encode('utf8')


	path = title.replace(" ","-")
	full_path = "bases/"+path+".txt"
	with open(full_path,"w+") as file:
		file.write(closed_news)

def commentsFromUser(data):
	_dict = dict()
	_dict['user_comment'] = data
	data_to_json = json.dumps(_dict, sort_keys=False,indent=4, ensure_ascii=False).encode('utf8')

	with open("userComments/comments.txt","a+") as file:
		file.write(data_to_json)


def getTupleCountries():
    tp = []
    with open("paises-gentilicos-google-maps.json", "r") as f:
        str = eval(f.read())
    for i in str:
        tp.append((i['sigla'],i['nome_pais']))
    
    return tp

def parse_state():
    tp = []
    with open("estados-cidades.json", "r") as f:
        str = json.loads(f.read())[0]
    for i in str['estados']:
        tp.append((i['sigla'],i['nome']))
    print (tp)







scrap.py
#encoding=utf-8
import json
from requests import get
from requests.exceptions import RequestException
from contextlib import closing
from bs4 import BeautifulSoup
import urllib2
import re

def simple_get(url):
    """
    Attempts to get the content at `url` by making an HTTP GET request.
    If the content-type of response is some kind of HTML/XML, return the
    text content, otherwise return None.
    """
    try:
        with closing(get(url, stream=True)) as resp:
            if is_good_response(resp):
                return resp.content
            else:
                return None

    except RequestException as e:
        log_error('Error during requests to {0} : {1}'.format(url, str(e)))
        return None

def is_good_response(resp):
    """
    Returns True if the response seems to be HTML, False otherwise.
    """
    content_type = resp.headers['Content-Type'].lower()
    return (resp.status_code == 200 
            and content_type is not None 
            and content_type.find('html') > -1)

def log_error(e):
    """
    It is always a good idea to log errors. 
    This function just prints them, but you can
    make it do anything.
    """
    print(e)

def getNewsLinks(url):
    html_page = urllib2.urlopen(url)
    soup = BeautifulSoup(html_page)


    for link in soup.findAll('a', attrs={'href': re.compile("^http://")}):
        print link.get('href')

def getTitle(url):
	raw_html = simple_get(url)
	html_paragraph =  BeautifulSoup(raw_html, 'html.parser')

	title = html_paragraph.select('title')
	return str(title[0]).split("|")[0].replace("<title>","")

def get_paragraph(url):
	#'https://exame.abril.com.br/economia/os-8-estados-brasileiros-que-devem-superar-o-nivel-pre-crise-em-2019'
	raw_html = simple_get(url)
	html_paragraph =  BeautifulSoup(raw_html, 'html.parser')

	str=""
	for p in html_paragraph.select('p'):
		#print(p.text)
		str= str+p.text+"</p><p>"

	return str


def getJsonReturnListOption(json):
    data = json
    
    list_d = []
    index_to_alpha = 0
    for each in data['geonames']:
        # print each['name'], each['fcode'], each['countryCode']
        iso_state = "/$$$"
        try:
            iso_state = each['adminName1']
            iso_state = '/'+iso_state
        except Exception as e:
            iso_state = "/$$$"
        if len(iso_state) == 0:
            iso_state = "/$$$"
        iso_state = iso_state.replace(' ','---')

        try:
            list_d.append(str(index_to_alpha) + " " + str(each['geonameId']) + " " + each['name'] + " " + each['fcode'] + " " +iso_state+" "+ each['countryCode'])
        except Exception as e:
            list_d.append(str(index_to_alpha) + " " + str(each['geonameId']) + " " + each['name'] + " " + each['fcode'] + " " +iso_state+ " .")
        
        index_to_alpha=index_to_alpha+1

    return list_d


def getJsonAndReturnListOption(json_data):
	data = json.loads(json_data)
	list_data = []
	for each in data['geonames']:
        
		list_data.append(each['name'] + " " + each['fcode'] )
        
	return list_data


def getToponymAndAttachJson(toponimos):

    quantity_results_wanted = '5'
    list_data_to_be_shown = []

    for each in toponimos:
        topoynim_without_spaces = each.replace(' ','+').lower()
        # http://api.geonames.org/searchJSON?formatted=true&q=rio+de+janeiro&maxRows=3&lang=es&username=artdcomp&style=short
        url_search = 'http://api.geonames.org/searchJSON?formatted=true&q='+topoynim_without_spaces+'&maxRows='+quantity_results_wanted+'&lang=pt&username=artdcomp&style=medium'
        json_info = get(url_search).json()
       
        parse_info_to_be_shown = getJsonReturnListOption(json_info)
        list_data_to_be_shown.append(parse_info_to_be_shown)

    return list_data_to_be_shown


















tutor.py
#!/usr/bin/env python
# -*- coding: utf-8 -*- 

from polyglot.text import Text, Word
import info_extract
import scrap

def lprint(s):
	for i in s:
		print i

def applicationDev(url):
	
	blob = scrap.get_paragraph(url)

	# todos os toponimos presentes no texto
	topon = info_extract.parse_text_toponimo(blob)

	# remove toponimos repetidos
	newlist_t=[ii for n,ii in enumerate(topon) if ii not in topon[:n]]

	# toponimos na ordem em que aparecem no texto
	toponimos = info_extract.sortToponymOccurence(info_extract.get_sentence(blob), newlist_t)

	# para cada toponimo -> pegar json associado e criar lista
	json_data = scrap.getToponymAndAttachJson(toponimos)
	
	#insere a tag select em cada toponimo
	select_options = info_extract.insertDataIntoSelectTag(json_data)

	# pega todas as frases da noticia
	sentence = info_extract.get_sentence(blob)

	# cria um botao com o nome do toponimo 
	toponym_button_text = info_extract.create_text_button(newlist_t)

	# texto gerado
	inicial_text = info_extract.replaceMultiple(sentence,newlist_t,toponym_button_text)
	
	# texto gerado
	text_with_button = info_extract.putButtonIndex(inicial_text)

	text_with_button_id = info_extract.putIndexInToponyms(text_with_button)
	
	final_text = info_extract.generateText(text_with_button_id, select_options)	

	return (final_text, toponimos)


flask_app.py
# coding=utf-8
import sys
reload(sys)
sys.setdefaultencoding('utf-8')
import scrap
import random
import files

from flask import request, Flask, render_template, url_for, redirect

app = Flask(__name__)

urls = files.getNewsDataFromUrls()

len_urls_available = len(urls)
toponyms = []
user_data = []
data_to_store = []
random_news = 0
title = ""
gen_data = []
noticia = ""
random_news_sorted = []
#=====

@app.route('/')
def index():
    return render_template("index.html")

@app.route('/avaliar', methods=['POST','GET'])
def avaliar():
	global toponyms
	global user_data
	global random_news
	global data_to_store
	global title
	global gen_data
	global noticia
	global random_news_sorted
	global aux_data_to_store_firebase
	global aux_file_name

	if request.method == 'POST':
		i=0
		qt_of_tops = len(toponyms)
	
		while qt_of_tops > 0:
			top = request.form.get('top_data_'+str(i))
			conf = request.form.get('confi_data_'+str(i))
			i = i+1
			qt_of_tops = qt_of_tops - 1
			if top is not None:
				if conf is not None:
					user_data.append((top,conf))

		data_to_store.append((urls[random_news][0], user_data))
		aux_data_to_store_firebase = files.dataFromUser(title, data_to_store)
		aux_file_name = title.replace(" ","_")
		
		# data_to_create_json = files.processData(title, urls[random_news][0])
		# #if news is closed then write into a file
		# if data_to_create_json != False:
		# 	files.generateNewsJsonFiles(data_to_create_json, toponyms, urls[random_news][0],title)

		
		del data_to_store[:]
		# limpa os dados, para nao acc 
		del user_data[:]

		if request.form['submit_button'] == 'Do Something':
			if len(random_news_sorted) == len(urls):
				del random_news_sorted[:]	            

			random_news = random.randint(0,(len_urls_available-1)) 
			while random_news in random_news_sorted:
				random_news = random.randint(0,(len_urls_available-1))
			random_news_sorted.append(random_news)

			url = urls[random_news][0]
			gen_data = urls[random_news][1]
			noticia = gen_data[0]
			toponyms = gen_data[1]
			title = scrap.getTitle(url)

			return render_template("avaliar.html", title=title, noticia=noticia, firebase=aux_data_to_store_firebase, firebase_filename=aux_file_name)

		elif request.form['submit_button'] == 'Do Something Else':
			return render_template("validar.html",firebase=aux_data_to_store_firebase, firebase_filename=aux_file_name)

	if request.method == 'GET':
		if len(random_news_sorted) == len(urls):
			del random_news_sorted[:]

		random_news = random.randint(0,(len_urls_available-1)) 
		while random_news in random_news_sorted:
			random_news = random.randint(0,(len_urls_available-1))
		random_news_sorted.append(random_news)

		url = urls[random_news][0]
		gen_data = urls[random_news][1]
		noticia = gen_data[0]
		toponyms = gen_data[1]
		title = scrap.getTitle(url)

		return render_template("avaliar.html", title=title, noticia=noticia)

@app.route('/validar', methods=['post','get'])
def validar():

	if request.method == "POST":
		user_data = request.form.get('comments')
		files.commentsFromUser(user_data)
		user_data = user_data+"$$$$$$$$$"
		return render_template('index.html',user_data=user_data)

	elif request.method == "GET":
		return render_template("validar.html")

if __name__ == '__main__':
	app.run(debug=True)


info_extract.py
# -*- coding: UTF-8 -*-
import sys
reload(sys)
sys.setdefaultencoding('utf8')
import re
import files

from polyglot.text import Text, Word

def parse_text_toponimo(blob_text):

	text = Text(blob_text)

	loc = ""
	data = []

	#gwt data
	for sent in text.sentences:
		#print(sent)
		for entity in sent.entities:
			#print(str(entity.tag))
			for x in entity:
				if len(entity) > 1:
					loc = loc+x+' '
				else:
					loc = x+' '

			if (loc.find(' - ') > 0):
				loc = loc.replace(' - ', '-')
			loc = loc[:-1]
			data.append((str(entity.tag),loc))
			loc = ""

	# encode to utf-8
	n_data = []
	for i in range(len(data)):
			n_data.append((data[i][0], data[i][1].encode("utf-8")))

	string_list = list()

	# select locations only
	for j in range(len(n_data)):
		if n_data[j][0] == "I-LOC":
			string_list.append(n_data[j][1])

	return string_list


def get_sentence(blob):
	text = Text(blob)
	s=""
	for sent in text.sentences:
		s=s+str(sent)+"\n"
	return s

def print_list(string_list):
	for st in string_list:
		print(st)


def specialCaseParseTopToUser(strr):

	s = strr
	codecs = [
				('PCLI', '(País'), 
				('ADM1', '(Estado'),
				('ADM2', '(Distrito do Estado'),
				('PPLC', '(Capital do País'),
				('PPLA', '(Capital do Estado'),
				('AIRP', '(Aeroporto'),
				('PPL' , '(Cidade'),
				('RGN' , '(Região'),
				('CONT', '(Continente'),
				('LCTY', '(Local Menor,Misto ou Indefinido'),
				('DAM', '(Barreira Dáqua'),
				('BLDG', '(Prédio'),
				('AGRF','(Instalação Agrícola'),
				('ADM3', '(Distrito da Cidade'),
				('SCH', '(Escola'),
				('LK', '(Lago'),
				('OPRA', '(Casa de Ópera'),
				('AMTH','(Anfiteatro'),
				('PAL', '(Palácio'),
				('STM','(Fluxo de Água'),
				('RVN','(Ravina'),
				('VLC','(Vulcão'),
				('ISL','(Ilha'),
				('BNK','(Banco'),
				('SHOL','(Cardume'),
				('HLL','(Colina'),
				('AREA','(Área'),
				('INLT','(Passagem'),
				('STDM','(Estádio'),
				('UNIV','(Universidade'),
				('PRK','(Parque'),
				('PCLH','(Antiga Entidade Política'),
				('RSV','(Reserva'),
				('SWMP','(Pântano'),
				('BSNU','(Bacia'),
				('MTS','(Montanhas'),
				('MT','(Montanha')
				]
	
	for i in codecs:
			if i[0] in strr:
				a1 = strr.split()[1]
				a2 = strr.split()[2]
				
				try:
					a3 = strr.split()[3]
					a4 = strr.split()[4]
					a5 = strr.split()[5]
				except:
					pass
				

				if i[0] == a1:
					s = strr.replace(strr.split()[1], i[1])
					s = s.replace('/', ' ')

				if i[0] == a2:
					s = strr.replace(strr.split()[2], i[1])
					s = s.replace('/', ' ')
				try:

					if i[0] == a3:
						s = strr.replace(a3, i[1])
						s= s.replace('/', ' ')

					if i[0] == a4:
						s = strr.replace(a4, i[1])
						s= s.replace('/', ' ')

					if i[0] == a5:
						s = strr.replace(a5, i[1])
						s= s.replace('/', ' ')
				except:
					pass

	return s

def parseTopToUser(strr):
	codecs = [
				('PCLI', '(País'), 
				('ADM1', '(Estado'),
				('ADM2', '(Distrito do Estado'),
				('PPLC', '(Capital do País'),
				('PPLA', '(Capital do Estado'),
				('AIRP', '(Aeroporto'),
				('PPL' , '(Cidade'),
				('RGN' , '(Região'),
				('CONT', '(Continente'),
				('LCTY', '(Local Menor,Misto ou Indefinido'),
				('DAM', '(Barreira Dáqua'),
				('BLDG', '(Prédio'),
				('AGRF','(Instalação Agrícola'),
				('ADM3', '(Distrito da Cidade'),
				('SCH', '(Escola'),
				('LK', '(Lago'),
				('OPRA', '(Casa de Ópera'),
				('AMTH','(Anfiteatro'),
				('PAL', '(Palácio'),
				('STM','(Fluxo de Água'),
				('RVN','(Ravina'),
				('VLC','(Vulcão'),
				('ISL','(Ilha'),
				('BNK','(Banco'),
				('SHOL','(Cardume'),
				('HLL','(Colina'),
				('AREA','(Área'),
				('INLT','(Passagem'),
				('STDM','(Estádio'),
				('UNIV','(Universidade'),
				('PRK','(Parque'),
				('PCLH','(Antiga Entidade Política'),
				('RSV','(Reserva'),
				('SWMP','(Pântano'),
				('BSNU','(Bacia'),
				('MTS','(Montanhas'),
				('MT','(Montanha'),
				('PCLD','(Entidade Política Dependente')
				]

	all_country = files.getTupleCountries()
	
	string = " ".join(strr.split()[2:])
	b = string
	c = string
	for i in codecs:
		if i[0] in string:
			b = string.replace(string.split()[-3],i[1])

	for i in all_country:
		if i[0] == b.split()[-1]:
			c = b.replace(b.split()[-1], '/ '+i[1])

	d = c.replace('---',' ')
	e = d.replace("$$$"," - ")
	e = e.replace('/ /', '/')
	e = e.replace('/)', ')')
	e = e.replace('.', ' ')

	return specialCaseParseTopToUser(e)+')'


# nao pega os toponimos mas sim o resultado da consuta nos json
def insertDataIntoSelectTag(data):
	data_list = []
	option_1_to_5 = '<option value="5"> 100 %</option> <option value="4"> 75 % </option> <option value="3"> 50 % </option> <option value="2"> 25 % </option> <option value="1"> 0 % </option><option value="None" selected>None</option>  </select>'
	index_quest = 0
	string_option = ""
	string_select = ""
	name_select_in_text = "top_data_"
	option_select_in_text = "confi_data_"


	option_none_of_alternatives_and_dont_know = '''<option value="5 0000000 Nenhuma das Alternativas">Nenhuma das Alternativas</option> <option value="6 0000000 Não é um lugar">Não é um lugar</option> <option value="7 0000000 Não Sei">Não Sei</option> <option value="8 0000000 -- None --" selected>-- None --</option>'''

	for i in range(len(data)):
		string_select = string_select + '<select class="form-control style_select" name="'+ name_select_in_text + str(i)+'" id="'+name_select_in_text + str(i)+'">'
		for j in data[i]:
			string_option = string_option+'<option value="'+j+'">' + parseTopToUser(j) + '</option>'
		
		string_select = string_select + string_option+ option_none_of_alternatives_and_dont_know + '</select>'
		data_list.append('''<div class="collapse" id="quest_''' + str(index_quest) + '''"> <div class="well"> <div class="row"> <div style="margin: 0 auto;" class="text-center"><div class="col-md-1">Topônimo:</div> <div class="col-md-6">''' + string_select +''' </div> <div class="col-md-2">  Certeza da Resposta :</div> <div class="col-md-1">'''+ '''<select class="form-control style_select" name="''' +option_select_in_text+str(index_quest)+ '''">'''+option_1_to_5 +''' </div> <div class="col-md-2" style="justify-items: center;"><button id="confirm_btn_id_''' + str(index_quest)+'''" class="btn btn-success style="margin-top:5px;" type="button" data-toggle="collapse" data-target="#quest_''' + str(index_quest) +''' " aria-expanded="false" aria-controls=quest_'''+ str(index_quest)+''' onclick="submitButtonStyle(this)" >Confirmar</button> ''' +  '''</div></div></div></div></div>''')
		index_quest = index_quest + 1
		
		string_option = ""
		string_select = ""
		
	return data_list


def putIndexInToponyms(text):
	spl  = text.split("<button")
	index_top = 0
	for i in range(len(spl) - 1):
		aux = 0
		if aux % 2 == 0:
			spl[i] = spl[i] + '<button id="toponym_in_news_' +str(index_top)+'"'
			index_top = index_top + 1
		aux = aux + 1

	return ''.join(spl)



def create_text_button(data):
	bt = []
	index = 0
	for i in data:# i remove aria-controls for test
		bt.append('''<button style="font-size:1em;"class="btn style_toponym_in_text main_form" type="button" data-toggle="collapse" aria-expanded="false" >''' + i + '''</button>''')
		index = index + 1
	return bt

def replaceMultiple(mainString, toBeReplaces, newString):
	# Iterate over the strings to be replaced    
	i = 0
	for elem in toBeReplaces :
		# Check if string is in the main string
		if elem in mainString :
			# Replace the string
			mainString = mainString.replace(elem, newString[i])
			i = i + 1
	return  mainString

def putButtonIndex(string):
	split_string =  string.split()
	index_toponimo = 0
	j = 0
	lista = []
	for i in split_string:
		if '<button' in i:
			i = i + ' data-target="#quest_'+ str(index_toponimo) +'" '
			lista.append((i,j))
			index_toponimo = index_toponimo + 1
		j = j + 1

	for i in range(len(lista)):
		split_string[lista[i][1]] = lista[i][0]

	return ' '.join(split_string)


def generate_paragraph(string):
	lista = string.split('\n')
	
	lista2 = []
	for x in range(len(lista)):
		lista2.append("<p>"+lista[x]+"</p>")

	sent = ""
	for x in lista2:
		sent = sent + x

	return sent


def keyToSort(item):
	return item[0]

def sortToponymOccurence(a,b):
	lista = []
	
	# encontra os indices de todos os toponimos no texto(na string)
	for x in b:
		indices_top = [m.start() for m in re.finditer(x, a)]
		lista.append((indices_top, x))
	
	tpn = []

	#add em tpn as repeticoes dos toponimos
	for k in lista:
		lenght = len(k[0])
		index_controle = 0
		while lenght>0:
			tpn.append((k[0][index_controle],k[1]))
			lenght = lenght - 1
			index_controle = index_controle + 1


	result = sorted(tpn, key=keyToSort)
	res = []
	for s in result:
		res.append(s[1])

	return res

# Create a function called "chunks" with two arguments, l and n:
def chunks(l, n):
	# For item i in a range that is a length of l,
	for i in range(0, len(l), n):
		# Create an index range for l of n items:
		yield l[i:i+n]

def createSelectGroup(select_options, qtde_blocks):
	lista = list(chunks(select_options, qtde_blocks))

	ls = []
	string = ""
	for i in lista:
		string = ""
		for j in i:
			string = string + j
		ls.append(string)

	return ls


def intercalateLists(l1, l2):

	size_l1 = len(l1)
	size_l2 = len(l2)

	final = []

	if size_l1 >= size_l2:
		for i in range(len(l2)):
			final.append(l1[i])
			final.append(l2[i])
			i = i + 2

		i = size_l2
		for i in range(size_l2,size_l1):
			final.append(l1[i])

	else: 
		for i in range(len(l1)):
			final.append(l1[i])
			final.append(l2[i])
			i = i + 2

		i = size_l1
		for i in range(size_l1, size_l2):
			final.append(l2[i])

	return final

def generateText(text, select_option):

	qtd_of_blocks = len(select_option)/3


	index_of_paragraphs = [m.start() for m in re.finditer("</p><p>", text)]

	tam_fixo = int (len(index_of_paragraphs) / qtd_of_blocks)
	tam = tam_fixo

	ll = []

	for k in range(len(index_of_paragraphs)):
		if k == tam:
			ll.append(index_of_paragraphs[k])
			tam = tam + tam_fixo

	leen = len(index_of_paragraphs)
	plus1 = index_of_paragraphs[leen-1]
	ll.append(plus1)

	l_aux_2 = map(lambda x: x + 4, ll)
	l_aux_2.insert(0, 0)
	

	listaa = []

	j=1
	for i in range(len(l_aux_2)-1):
		listaa.append(text[l_aux_2[i]:l_aux_2[j]])
		j = j+1

	select_block_size = int(len(select_option)/qtd_of_blocks)
	new_select = createSelectGroup(select_option, select_block_size)
	final_text = intercalateLists(new_select,listaa)
	
	string = ""
	for i in final_text:
		string = string + i
		
	return string
